packages/firebase_messaging/android/src/main/java/io/flutter/plugins/firebasemessaging/FlutterFirebaseMessagingService.java

```diff
@@ -42,4 +128,148 @@ public void onNewToken(String token) {
+    }
+  }
+
+  public static void setBackgroundChannel(MethodChannel channel) {
+    sBackgroundChannel = channel;
+  }
+
+  public static void setPluginRegistrant(PluginRegistry.PluginRegistrantCallback callback) {
+    sPluginRegistrantCallback = callback;
+  }
+
+  public static void setBackgroundMessageHandle(Context context, Long handle) {
+    sBackgroundMessageHandle = handle;
+
+    // Store background message handle in shared preferences so it can be retrieved
+    // by other application instances.
+    SharedPreferences prefs = context.getSharedPreferences(SHARED_PREFERENCES_KEY, 0);
+    prefs.edit().putLong(BACKGROUND_MESSAGE_CALLBACK_HANDLE_KEY, handle).apply();
+  }
+
+  public static void setBackgroundSetupHandle(Context context, long callbackHandle) {
+    // Store background setup handle in shared preferences so it can be retrieved
+    // by other application instances.
+    SharedPreferences prefs = context.getSharedPreferences(SHARED_PREFERENCES_KEY, 0);
+    prefs.edit().putLong(BACKGROUND_SETUP_CALLBACK_HANDLE_KEY, callbackHandle).apply();
+  }
+
+  public static Long getOnMessageCallbackHandle(Context context) {
+    return context
+        .getSharedPreferences(SHARED_PREFERENCES_KEY, 0)
+        .getLong(BACKGROUND_MESSAGE_CALLBACK_HANDLE_KEY, 0);
+  }
+
+  private static void executeDartCallbackInBackgroundIsolate(
+      Context context, RemoteMessage remoteMessage, final CountDownLatch latch) {
+    if (sBackgroundChannel == null) {
+      throw new RuntimeException(
+          "setBackgroundChannel was not called before messages came in, exiting.");
+    }
+
+    // If another thread is waiting, then wake that thread when the callback returns a result.
+    MethodChannel.Result result = null;
+    if (latch != null) {
+      result =
+          new MethodChannel.Result() {
+            @Override
+            public void success(Object result) {
+              latch.countDown();
+            }
+
+            @Override
+            public void error(String errorCode, String errorMessage, Object errorDetails) {
+              latch.countDown();
+            }
+
+            @Override
+            public void notImplemented() {
+              latch.countDown();
+            }
+          };
+    }
+
+    Map<String, Object> args = new HashMap<>();
+    Map<String, Object> messageData = new HashMap<>();
+    if (sBackgroundMessageHandle == null) {
+      sBackgroundMessageHandle = getOnMessageCallbackHandle(context);
+    }
+    args.put("handle", sBackgroundMessageHandle);
+
+    if (remoteMessage.getData() != null) {
+      messageData.put("data", remoteMessage.getData());
+    }
+    if (remoteMessage.getNotification() != null) {
+      messageData.put("notification", remoteMessage.getNotification());
+    }
+
+    args.put("message", messageData);
+
+    sBackgroundChannel.invokeMethod("", args, result);
+  }
+
+  // TODO(kroikie): Find a better way to determine application state.
```

Comment:
@matthew-carroll might have ideas for other ways to do this

Reply:
@kroikie can you clarify what you mean technically by "foreground" in this case?

Reply:
@matthew-carroll the goal here is to determine when the Flutter process is still alive, in which case the original isolate can be used to process incoming messages. However when the Flutter process is no longer running the "background" isolate is used to process incoming messages.

Here we are checking if the app is in the foreground and visible, which is not exactly what the goal is but I could not find a clean way to determine if the Flutter process is still alive, this approximation was the closest I could get.

Reply:
There should only be 1 process, and that's the app's process. If this code is executing then that process exists, by definition...

Reply:
@matthew-carroll I'm not sure if this is actionable feedback for Arthur, are you ok with landing this as is?